services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestHandler.java

```diff
@@ -383,7 +383,18 @@ private int getSubPartitionId(
   }
 
   private PerStoreVersionState getPerStoreVersionState(String storeVersion) {
-    return perStoreVersionStateMap.computeIfAbsent(storeVersion, this::generatePerStoreVersionState);
+    return perStoreVersionStateMap.compute(storeVersion, (ignored, versionState) -> {
+      if (versionState == null) {
+        return this.generatePerStoreVersionState(storeVersion);
+      } else {
+        /**
+         * Try to refresh the storage engine reference for every request in case {@link com.linkedin.davinci.store.StorageEngineFactory}
+         * recreates a new instance during re-balance, which may drop the storage engine and recreate a new one.
+         */
+        versionState.storageEngine = storageEngineRepository.getLocalStorageEngine(storeVersion);
```

Comment:
If we're going to look this up every time on the hot path, then I don't think it makes sense to have it in the cached mapping... let's just remove it entirely from the `PerStoreVersionState` object and look it up separately.

Or alternatively, we don't look it up in the hot path, but the Helix state transition handling code can call a hook to update and/or clear this mapping when dealing with the first/last partition... that is likely more complicated though.

Reply:
There are certainly ways to make it more efficient.
Yesterday, there was GCN pressure, so we made a simple solution.

In theory, we can check whether the storage engine is empty or not and based on the current logic, we only need to refresh the reference when the storage engine instance kept in `PerStoreVersionState` is empty.

Reply:
Yeah, I understand the need for a tactical fix. Thanks a lot for putting it together, and my apologies for introducing the regression.

My only concern with the fix in this PR is that the closure passed into the `CHM`'s `compute` function is guaranteed to be executed atomically, which seems to me like it could be a point of high contention, especially in dedicated clusters containing one or few stores that all have a high throughput...

I suggest an alternative approach in this PR, which I believe may be more efficient: https://github.com/linkedin/venice/pull/663

Thanks again for the late night hot fix.